File: legacy-node/src/chunker.js
import { sha256Text } from "./hash.js";

export function chunkMarkdown(markdown, options = {}) {
  const targetTokens = options.targetTokens || 420;
  const maxTokens = options.maxTokens || 620;
  const sections = splitByHeadings(markdown);
  const chunks = [];
  let ordinal = 0;

  for (const section of sections) {
    const paragraphs = section.text.split(/\n{2,}/).map((p) => p.trim()).filter(Boolean);
    let buffer = [];
    let bufferTokens = 0;
    for (const paragraph of paragraphs) {
      const tokens = estimateTokens(paragraph);
      if (buffer.length && bufferTokens + tokens > maxTokens) {
        chunks.push(makeChunk(buffer.join("\n\n"), section.heading, ordinal++));
        buffer = [];
        bufferTokens = 0;
      }
      if (tokens > maxTokens) {
        for (const piece of splitLargeParagraph(paragraph, targetTokens)) {
          chunks.push(makeChunk(piece, section.heading, ordinal++));
        }
      } else {
        buffer.push(paragraph);
        bufferTokens += tokens;
      }
    }
    if (buffer.length) chunks.push(makeChunk(buffer.join("\n\n"), section.heading, ordinal++));
  }

  return chunks;
}

export function estimateTokens(text) {
  return Math.max(1, Math.ceil(text.length / 4));
}

function splitByHeadings(markdown) {
  const lines = markdown.split("\n");
  const sections = [];
  let current = { heading: undefined, text: "" };
  for (const line of lines) {
    const heading = /^(#{1,6})\s+(.+)$/.exec(line);
    if (heading && current.text.trim()) {
      sections.push(current);
      current = { heading: heading[2].trim(), text: `${line}\n` };
    } else {
      if (heading) current.heading = heading[2].trim();
      current.text += `${line}\n`;
    }
  }
  if (current.text.trim()) sections.push(current);
  return sections;
}

function splitLargeParagraph(text, targetTokens) {
  const words = text.split(/\s+/);
  const targetWords = Math.max(60, Math.floor(targetTokens * 0.75));
  const pieces = [];
  for (let i = 0; i < words.length; i += targetWords) {
    pieces.push(words.slice(i, i + targetWords).join(" "));
  }
  return pieces;
}

function makeChunk(text, heading, ordinal) {
  const clean = text.trim();
  return {
    id: sha256Text(`${ordinal}:${clean}`).slice(0, 24),
    ordinal,
    text: clean,
    heading,
    token_count: estimateTokens(clean),
    hash: sha256Text(clean)
  };
}

File: rust/chunker.rs
use crate::hash::hash_text;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Chunk {
    pub id: String,
    pub ordinal: usize,
    pub text: String,
    pub heading: Option<String>,
    pub page: Option<i64>,
    pub slide: Option<i64>,
    pub token_count: usize,
    pub hash: String,
}

pub fn estimate_tokens(text: &str) -> usize {
    std::cmp::max(1, text.chars().count().div_ceil(4))
}

pub fn chunk_markdown(markdown: &str) -> Vec<Chunk> {
    let sections = split_by_headings(markdown);
    let mut chunks = Vec::new();
    let mut ordinal = 0usize;

    for (heading, text) in sections {
        let paragraphs: Vec<_> = text
            .split("\n\n")
            .map(str::trim)
            .filter(|p| !p.is_empty())
            .collect();
        let mut buffer: Vec<String> = Vec::new();
        let mut buffer_tokens = 0usize;
        for paragraph in paragraphs {
            let tokens = estimate_tokens(paragraph);
            if !buffer.is_empty() && buffer_tokens + tokens > 620 {
                chunks.push(make_chunk(&buffer.join("\n\n"), heading.clone(), ordinal));
                ordinal += 1;
                buffer.clear();
                buffer_tokens = 0;
            }
            buffer.push(paragraph.to_string());
            buffer_tokens += tokens;
        }
        if !buffer.is_empty() {
            chunks.push(make_chunk(&buffer.join("\n\n"), heading, ordinal));
            ordinal += 1;
        }
    }
    chunks
}

fn split_by_headings(markdown: &str) -> Vec<(Option<String>, String)> {
    let mut sections = Vec::new();
    let mut heading = None;
    let mut text = String::new();
    for line in markdown.lines() {
        if line.starts_with('#') && line.contains(' ') && !text.trim().is_empty() {
            sections.push((heading.clone(), text.clone()));
            heading = line.split_once(' ').map(|(_, h)| h.trim().to_string());
            text.clear();
        } else if line.starts_with('#') && line.contains(' ') {
            heading = line.split_once(' ').map(|(_, h)| h.trim().to_string());
        }
        text.push_str(line);
        text.push('\n');
    }
    if !text.trim().is_empty() {
        sections.push((heading, text));
    }
    sections
}

fn make_chunk(text: &str, heading: Option<String>, ordinal: usize) -> Chunk {
    let clean = text.trim().to_string();
    Chunk {
        id: hash_text(&format!("{ordinal}:{clean}"))[..24].to_string(),
        ordinal,
        token_count: estimate_tokens(&clean),
        hash: hash_text(&clean),
        text: clean,
        heading,
        page: None,
        slide: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chunks_markdown_by_headings_and_preserves_text() {
        let chunks = chunk_markdown("# Title\n\nAlpha beta gamma.\n\n## Next\n\nDelta epsilon.");
        assert_eq!(chunks.len(), 2);
        assert_eq!(chunks[0].heading.as_deref(), Some("Title"));
        assert!(chunks[1].text.contains("Delta"));
    }
}
